runtimes/dart-2.18/server.dart

```diff
@@ -10,56 +10,121 @@ import 'function_types.dart';
 
 void main() async {
   await shelf_io.serve((req) async {
-    List<String> userLogs = [];
-    if (req.method != 'POST') {
-      return shelf.Response(500, body: 'Invalid request');
+    if ((req.headers['x-open-runtimes-secret'] ?? '') == '' || (req.headers['x-open-runtimes-secret'] ?? '') != (Platform.environment['OPEN_RUNTIMES_SECRET'] ?? '')) {
+      return shelf.Response(500, body: 'Unauthorized. Provide correct "x-open-runtimes-secret" header.');
     }
-    if (req.headers['x-internal-challenge'] !=
-        Platform.environment['INTERNAL_RUNTIME_KEY']) {
-      return shelf.Response(500, body: 'Unauthorized');
+
+    String bodyString = await req.readAsString();
+    dynamic body = bodyString;
+    String method = req.method;
+    Map<String, dynamic> headers = {};
+
+    for (MapEntry entry in req.headers.entries) {
+      String header = entry.key.toLowerCase();
+      if(!header.startsWith('x-open-runtimes-')) {
+        headers[header] = entry.value;
+      }
+    }
+
+    String contentType = req.headers['content-type'] ?? 'plain/text';
+    if(contentType.contains('application/json')) {
+      if(!bodyString.isEmpty) {
+        body = jsonDecode(bodyString);
+      } else {
+        body = {};
+      }
+    }
+
+    String hostHeader = req.headers['host'] ?? '';
+
+    String path = '/' + req.url.path;
+    String host = '';
+    String scheme = req.headers['x-forwarded-proto'] ?? 'http';
+    String defaultPort = scheme == 'https' ? '443' : '80';
+    int port = int.parse(defaultPort);
+    String queryString = req.url.query;
+    Map<String, String> query = {};
+
+    if(hostHeader.contains(':')) {
+      host = hostHeader.split(':')[0];
+      port = int.parse(hostHeader.split(':')[1]);
+    } else {
+      host = hostHeader;
+      port = int.parse(defaultPort);
+    }
+
+    for(final param in queryString.split("&")) {
+      final parts = param.split("=");
+
+      final key = parts[0];
+      final value = parts.sublist(1).join('=');
+
+      if(key != null && !key.isEmpty) {
+        query[key] = value;
+      }
+    }
```

Comment:
I don't think we should manually parse the query string like this because we might not catch some edge cases with & and =. Also, this doesn't quite handle arrays well.

Reply:
Hmm we currently dont support arrays (`?key[]=a&key[]=b`), as well as json-like structure (`?key[name]=a&key[role]=b`).
Tho this could be problematic in some strictily-typed languages, as value value be `string`, `string[]`, or `{ string: string }`

Reply:
Also not sure if it's possible to entirely avoid this in all runtimes. I would imagine that C++ doesnt have a native way to do this, and would require a library, or a solution just this one.